{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install scanpy\n",
    "!pip install KDEpy\n",
    "!pip install leidenalg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install -e ../../tools/nomad/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from datetime import datetime\n",
    "\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.sparse as sps\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "# to be able to import from tools dir\n",
    "module_path = os.path.abspath(os.path.join(\"../../\"))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import tools.util_probe as up\n",
    "import tools.util as ut\n",
    "import tools.NB_est as nb\n",
    "import tools.countsplit as cs\n",
    "import tools.scDEED as scd\n",
    "import tools.clustering_opt as co\n",
    "import tools.util_plot as nmd_plot\n",
    "import tools.util_imputation as nmd_imp\n",
    "\n",
    "import fi_nomad as nmd\n",
    "from fi_nomad.types import kernelInputTypes\n",
    "from fi_nomad.types import KernelStrategy\n",
    "from fi_nomad.types import InitializationStrategy\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from scipy.stats import kendalltau\n",
    "import importlib\n",
    "import logging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# logging has to manually turned on to see nomad output\n",
    "logging.basicConfig(level=logging.INFO)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = \"../../data/S2\"\n",
    "figure_path = f\"{data_path}/figures/imputation\"\n",
    "layer = \"counts\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1255 × 5540\n",
       "    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'outlier', 'n_genes', 'total_counts_norm', 'total_counts_scale', 'embedding_reliability', 'reliability_score', 'null_reliability_score', 'leiden_opt_PCA'\n",
       "    var: 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'var_counts', 'is_scd_outlier', 'nb_overdisp', 'nb_overdisp_cutoff', 'nb_mean', 'nb_umi', 'Intercept_step1_sct', 'log_umi_step1_sct', 'dispersion_step1_sct', 'mean', 'std'\n",
       "    uns: 'BacSC_params_PCA', 'PCA', 'leiden', 'leiden_opt_PCA_colors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'normalized_counts', 'vst_counts'\n",
       "    obsp: 'PCA_connectivities', 'PCA_distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_counts = sc.read_h5ad(data_path + \"/filtered_data_maxpool_processed_and_pca.h5ad\")\n",
    "data_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read expression level class data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>tpm</th>\n",
       "      <th>expression_level</th>\n",
       "      <th>log_tpm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA0001</th>\n",
       "      <td>dnaA</td>\n",
       "      <td>333.097801</td>\n",
       "      <td>high</td>\n",
       "      <td>8.379802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0002</th>\n",
       "      <td>dnaN</td>\n",
       "      <td>299.127321</td>\n",
       "      <td>high</td>\n",
       "      <td>8.224616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0003</th>\n",
       "      <td>recF</td>\n",
       "      <td>128.151975</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.001712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0004</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>138.579052</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.114565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0005</th>\n",
       "      <td>lptA</td>\n",
       "      <td>141.413217</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.143773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5566</th>\n",
       "      <td>PA5566</td>\n",
       "      <td>45.077831</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.494346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5567</th>\n",
       "      <td>PA5567</td>\n",
       "      <td>49.149167</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.619095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5568</th>\n",
       "      <td>PA5568</td>\n",
       "      <td>208.178186</td>\n",
       "      <td>high</td>\n",
       "      <td>7.701675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5569</th>\n",
       "      <td>rnpA</td>\n",
       "      <td>3387.343294</td>\n",
       "      <td>high</td>\n",
       "      <td>11.725938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5570</th>\n",
       "      <td>rpmH</td>\n",
       "      <td>4824.078870</td>\n",
       "      <td>high</td>\n",
       "      <td>12.236038</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Name          tpm expression_level    log_tpm\n",
       "locus_tag                                                 \n",
       "PA0001       dnaA   333.097801             high   8.379802\n",
       "PA0002       dnaN   299.127321             high   8.224616\n",
       "PA0003       recF   128.151975      medium_high   7.001712\n",
       "PA0004       gyrB   138.579052      medium_high   7.114565\n",
       "PA0005       lptA   141.413217      medium_high   7.143773\n",
       "...           ...          ...              ...        ...\n",
       "PA5566     PA5566    45.077831       medium_low   5.494346\n",
       "PA5567     PA5567    49.149167       medium_low   5.619095\n",
       "PA5568     PA5568   208.178186             high   7.701675\n",
       "PA5569       rnpA  3387.343294             high  11.725938\n",
       "PA5570       rpmH  4824.078870             high  12.236038\n",
       "\n",
       "[5572 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_class = pd.read_csv(\n",
    "    \"../../data/control_expression_classification.tsv\", delimiter=\"\\t\"\n",
    ")\n",
    "expr_class.set_index(keys=\"locus_tag\", inplace=True)\n",
    "expr_class"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- `mean_counts`: mean gene expression\n",
    "- `n_cells`: number of cells that express gene (so (n - `n_cells`) / n = sparsity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>log1p_mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>var_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>nb_umi</th>\n",
       "      <th>Intercept_step1_sct</th>\n",
       "      <th>log_umi_step1_sct</th>\n",
       "      <th>dispersion_step1_sct</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>Name</th>\n",
       "      <th>tpm</th>\n",
       "      <th>expression_level</th>\n",
       "      <th>log_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA0001</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>508</td>\n",
       "      <td>0.529880</td>\n",
       "      <td>0.425190</td>\n",
       "      <td>59.521912</td>\n",
       "      <td>665.0</td>\n",
       "      <td>6.501290</td>\n",
       "      <td>508</td>\n",
       "      <td>0.555083</td>\n",
       "      <td>...</td>\n",
       "      <td>2.142832</td>\n",
       "      <td>-6.610836</td>\n",
       "      <td>2.011342</td>\n",
       "      <td>0.009302</td>\n",
       "      <td>1.105347</td>\n",
       "      <td>0.166192</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>333.097801</td>\n",
       "      <td>high</td>\n",
       "      <td>8.379802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0002</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>383</td>\n",
       "      <td>0.399203</td>\n",
       "      <td>0.335903</td>\n",
       "      <td>69.482072</td>\n",
       "      <td>501.0</td>\n",
       "      <td>6.218600</td>\n",
       "      <td>383</td>\n",
       "      <td>0.491633</td>\n",
       "      <td>...</td>\n",
       "      <td>2.263982</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.770695</td>\n",
       "      <td>0.190039</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>299.127321</td>\n",
       "      <td>high</td>\n",
       "      <td>8.224616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0003</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>604</td>\n",
       "      <td>0.728287</td>\n",
       "      <td>0.547131</td>\n",
       "      <td>51.872510</td>\n",
       "      <td>914.0</td>\n",
       "      <td>6.818924</td>\n",
       "      <td>604</td>\n",
       "      <td>1.023383</td>\n",
       "      <td>...</td>\n",
       "      <td>2.035867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.244879</td>\n",
       "      <td>0.176504</td>\n",
       "      <td>recF</td>\n",
       "      <td>128.151975</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.001712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0004</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>310</td>\n",
       "      <td>0.311554</td>\n",
       "      <td>0.271213</td>\n",
       "      <td>75.298805</td>\n",
       "      <td>391.0</td>\n",
       "      <td>5.971262</td>\n",
       "      <td>310</td>\n",
       "      <td>0.372257</td>\n",
       "      <td>...</td>\n",
       "      <td>2.366935</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.465868</td>\n",
       "      <td>0.219582</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>138.579052</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.114565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0005</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>108</td>\n",
       "      <td>0.103586</td>\n",
       "      <td>0.098565</td>\n",
       "      <td>91.394422</td>\n",
       "      <td>130.0</td>\n",
       "      <td>4.875197</td>\n",
       "      <td>108</td>\n",
       "      <td>0.131103</td>\n",
       "      <td>...</td>\n",
       "      <td>2.733139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.030973</td>\n",
       "      <td>0.314385</td>\n",
       "      <td>lptA</td>\n",
       "      <td>141.413217</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.143773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5566</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>6</td>\n",
       "      <td>0.004781</td>\n",
       "      <td>0.004769</td>\n",
       "      <td>99.521912</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.945910</td>\n",
       "      <td>6</td>\n",
       "      <td>0.004758</td>\n",
       "      <td>...</td>\n",
       "      <td>3.211789</td>\n",
       "      <td>-15.931712</td>\n",
       "      <td>3.485019</td>\n",
       "      <td>0.010852</td>\n",
       "      <td>-4.151651</td>\n",
       "      <td>0.232698</td>\n",
       "      <td>PA5566</td>\n",
       "      <td>45.077831</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.494346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5567</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>127</td>\n",
       "      <td>0.117928</td>\n",
       "      <td>0.111477</td>\n",
       "      <td>89.880478</td>\n",
       "      <td>148.0</td>\n",
       "      <td>5.003946</td>\n",
       "      <td>127</td>\n",
       "      <td>0.137487</td>\n",
       "      <td>...</td>\n",
       "      <td>2.692861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.840724</td>\n",
       "      <td>0.317603</td>\n",
       "      <td>PA5567</td>\n",
       "      <td>49.149167</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.619095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5568</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>466</td>\n",
       "      <td>0.506773</td>\n",
       "      <td>0.409970</td>\n",
       "      <td>62.868526</td>\n",
       "      <td>636.0</td>\n",
       "      <td>6.456770</td>\n",
       "      <td>466</td>\n",
       "      <td>0.598958</td>\n",
       "      <td>...</td>\n",
       "      <td>2.168977</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.040103</td>\n",
       "      <td>0.171432</td>\n",
       "      <td>PA5568</td>\n",
       "      <td>208.178186</td>\n",
       "      <td>high</td>\n",
       "      <td>7.701675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5569</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>1104</td>\n",
       "      <td>2.755378</td>\n",
       "      <td>1.323189</td>\n",
       "      <td>12.031873</td>\n",
       "      <td>3458.0</td>\n",
       "      <td>8.148735</td>\n",
       "      <td>1104</td>\n",
       "      <td>8.395140</td>\n",
       "      <td>...</td>\n",
       "      <td>1.816896</td>\n",
       "      <td>-5.261660</td>\n",
       "      <td>2.109112</td>\n",
       "      <td>0.094811</td>\n",
       "      <td>1.075991</td>\n",
       "      <td>0.444127</td>\n",
       "      <td>rnpA</td>\n",
       "      <td>3387.343294</td>\n",
       "      <td>high</td>\n",
       "      <td>11.725938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5570</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>292</td>\n",
       "      <td>0.321912</td>\n",
       "      <td>0.279079</td>\n",
       "      <td>76.733068</td>\n",
       "      <td>404.0</td>\n",
       "      <td>6.003887</td>\n",
       "      <td>292</td>\n",
       "      <td>0.463703</td>\n",
       "      <td>...</td>\n",
       "      <td>2.371253</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.460475</td>\n",
       "      <td>0.241673</td>\n",
       "      <td>rpmH</td>\n",
       "      <td>4824.078870</td>\n",
       "      <td>high</td>\n",
       "      <td>12.236038</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5540 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          feature_types genome  n_cells_by_counts  mean_counts  \\\n",
       "PA0001  Gene Expression   PA01                508     0.529880   \n",
       "PA0002  Gene Expression   PA01                383     0.399203   \n",
       "PA0003  Gene Expression   PA01                604     0.728287   \n",
       "PA0004  Gene Expression   PA01                310     0.311554   \n",
       "PA0005  Gene Expression   PA01                108     0.103586   \n",
       "...                 ...    ...                ...          ...   \n",
       "PA5566  Gene Expression   PA01                  6     0.004781   \n",
       "PA5567  Gene Expression   PA01                127     0.117928   \n",
       "PA5568  Gene Expression   PA01                466     0.506773   \n",
       "PA5569  Gene Expression   PA01               1104     2.755378   \n",
       "PA5570  Gene Expression   PA01                292     0.321912   \n",
       "\n",
       "        log1p_mean_counts  pct_dropout_by_counts  total_counts  \\\n",
       "PA0001           0.425190              59.521912         665.0   \n",
       "PA0002           0.335903              69.482072         501.0   \n",
       "PA0003           0.547131              51.872510         914.0   \n",
       "PA0004           0.271213              75.298805         391.0   \n",
       "PA0005           0.098565              91.394422         130.0   \n",
       "...                   ...                    ...           ...   \n",
       "PA5566           0.004769              99.521912           6.0   \n",
       "PA5567           0.111477              89.880478         148.0   \n",
       "PA5568           0.409970              62.868526         636.0   \n",
       "PA5569           1.323189              12.031873        3458.0   \n",
       "PA5570           0.279079              76.733068         404.0   \n",
       "\n",
       "        log1p_total_counts  n_cells  var_counts  ...    nb_umi  \\\n",
       "PA0001            6.501290      508    0.555083  ...  2.142832   \n",
       "PA0002            6.218600      383    0.491633  ...  2.263982   \n",
       "PA0003            6.818924      604    1.023383  ...  2.035867   \n",
       "PA0004            5.971262      310    0.372257  ...  2.366935   \n",
       "PA0005            4.875197      108    0.131103  ...  2.733139   \n",
       "...                    ...      ...         ...  ...       ...   \n",
       "PA5566            1.945910        6    0.004758  ...  3.211789   \n",
       "PA5567            5.003946      127    0.137487  ...  2.692861   \n",
       "PA5568            6.456770      466    0.598958  ...  2.168977   \n",
       "PA5569            8.148735     1104    8.395140  ...  1.816896   \n",
       "PA5570            6.003887      292    0.463703  ...  2.371253   \n",
       "\n",
       "        Intercept_step1_sct  log_umi_step1_sct  dispersion_step1_sct  \\\n",
       "PA0001            -6.610836           2.011342              0.009302   \n",
       "PA0002                  NaN                NaN                   NaN   \n",
       "PA0003                  NaN                NaN                   NaN   \n",
       "PA0004                  NaN                NaN                   NaN   \n",
       "PA0005                  NaN                NaN                   NaN   \n",
       "...                     ...                ...                   ...   \n",
       "PA5566           -15.931712           3.485019              0.010852   \n",
       "PA5567                  NaN                NaN                   NaN   \n",
       "PA5568                  NaN                NaN                   NaN   \n",
       "PA5569            -5.261660           2.109112              0.094811   \n",
       "PA5570                  NaN                NaN                   NaN   \n",
       "\n",
       "            mean       std    Name          tpm  expression_level    log_tpm  \n",
       "PA0001  1.105347  0.166192    dnaA   333.097801              high   8.379802  \n",
       "PA0002  0.770695  0.190039    dnaN   299.127321              high   8.224616  \n",
       "PA0003  1.244879  0.176504    recF   128.151975       medium_high   7.001712  \n",
       "PA0004  0.465868  0.219582    gyrB   138.579052       medium_high   7.114565  \n",
       "PA0005 -1.030973  0.314385    lptA   141.413217       medium_high   7.143773  \n",
       "...          ...       ...     ...          ...               ...        ...  \n",
       "PA5566 -4.151651  0.232698  PA5566    45.077831        medium_low   5.494346  \n",
       "PA5567 -0.840724  0.317603  PA5567    49.149167        medium_low   5.619095  \n",
       "PA5568  1.040103  0.171432  PA5568   208.178186              high   7.701675  \n",
       "PA5569  1.075991  0.444127    rnpA  3387.343294              high  11.725938  \n",
       "PA5570  0.460475  0.241673    rpmH  4824.078870              high  12.236038  \n",
       "\n",
       "[5540 rows x 24 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_counts.var = data_counts.var.join(expr_class)\n",
    "data_counts.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit NMD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "latent_rank = 10\n",
    "momentum_beta = 0.7\n",
    "max_iterations = 1000\n",
    "\n",
    "init_strat = InitializationStrategy.ROWWISE_MEAN\n",
    "kernel_strat = KernelStrategy.MOMENTUM_3_BLOCK_MODEL_FREE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "m, n = data_counts.X.shape\n",
    "X_nmd = ut.convert_to_dense(data_counts, layer=layer)\n",
    "X_nmd = X_nmd.astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:fi_nomad.entry:\tInitiating run, target_rank: 10, tolerance: None\n",
      "INFO:fi_nomad.entry:1000 total, final loss Not Tracked\n",
      "INFO:fi_nomad.entry:\tInitialization took 19.044169648026582 loop took 185.3114944039844 overall (0.1853114944039844/ea)\n"
     ]
    }
   ],
   "source": [
    "result = nmd.decompose(\n",
    "    X_nmd,\n",
    "    latent_rank,\n",
    "    kernel_strategy=kernel_strat,\n",
    "    initialization=init_strat,\n",
    "    kernel_params=kernelInputTypes.Momentum3BlockAdditionalParameters(\n",
    "        momentum_beta=momentum_beta\n",
    "    ),\n",
    "    manual_max_iterations=max_iterations,\n",
    "    verbose=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_counts.obsm[\"X_nmd\"] = result.factors[0]\n",
    "data_counts.varm[\"NMD_components\"] = result.factors[1].T\n",
    "data_counts.layers[\"Theta\"] = result.factors[0] @ result.factors[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output file for pretty plotting in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(data_counts.layers[\"Theta\"]).to_csv(\n",
    "    f\"{data_path}/Theta_3BNMD_r{latent_rank}.csv\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample gene pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'tools.util_imputation' from '/Users/stffn/projects/thesis_dev/msc_thesis/application/tools/util_imputation.py'>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(nmd_imp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes = 50  # number of gene pairs\n",
    "expression_levels_2_compare = [\"low\", \"medium_high\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>low</th>\n",
       "      <th>medium_high</th>\n",
       "      <th>dist</th>\n",
       "      <th>gene_pair_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA0570</td>\n",
       "      <td>PA1139</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA2018</td>\n",
       "      <td>PA1651</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA2599</td>\n",
       "      <td>PA2489</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA0825</td>\n",
       "      <td>PA4108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA2552</td>\n",
       "      <td>PA1915</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      low medium_high      dist  gene_pair_id\n",
       "0  PA0570      PA1139  0.000026             0\n",
       "1  PA2018      PA1651  0.000000             1\n",
       "2  PA2599      PA2489  0.000000             2\n",
       "3  PA0825      PA4108  0.000000             3\n",
       "4  PA2552      PA1915  0.000000             4"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pairs = nmd_imp.sample_gene_pairs(\n",
    "    data_counts,\n",
    "    n=n_genes,\n",
    "    measure_var=[\"mean_counts\", \"pct_dropout_by_counts\"],\n",
    "    expr_lvls=expression_levels_2_compare,\n",
    ")\n",
    "gene_pairs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get all negative Theta values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "      <th>gene_pair_id</th>\n",
       "      <th>dist</th>\n",
       "      <th>expression_lvl_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA0570</td>\n",
       "      <td>0.043325</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA0570</td>\n",
       "      <td>0.065547</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA0570</td>\n",
       "      <td>0.039778</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA0570</td>\n",
       "      <td>0.019333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA0570</td>\n",
       "      <td>0.020767</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene  Expression  gene_pair_id      dist expression_lvl_class\n",
       "0  PA0570    0.043325             0  0.000026                  low\n",
       "1  PA0570    0.065547             0  0.000026                  low\n",
       "2  PA0570    0.039778             0  0.000026                  low\n",
       "3  PA0570    0.019333             0  0.000026                  low\n",
       "4  PA0570    0.020767             0  0.000026                  low"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_df = nmd_imp.make_gene_pair_df(gene_pairs, adata=data_counts)\n",
    "gene_pair_df.sort_values(by=[\"gene_pair_id\", \"dist\"]).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export to csv for visualization in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_df.to_csv(\n",
    "    f\"{data_path}/gene_pair_df_4_viz_{n_genes}genes_3BNMD_rank{latent_rank}_{expression_levels_2_compare[0]}-{expression_levels_2_compare[1]}_mean_and_sparsity_whereXzero.csv\",\n",
    "    index=False,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bacsc-jupyter-lrz",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
