{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install scanpy\n",
    "!pip install KDEpy\n",
    "!pip install leidenalg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install -e ../../tools/nomad/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from datetime import datetime\n",
    "\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.sparse as sps\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "# to be able to import from tools dir\n",
    "module_path = os.path.abspath(os.path.join(\"../../\"))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import tools.util_probe as up\n",
    "import tools.util as ut\n",
    "import tools.NB_est as nb\n",
    "import tools.countsplit as cs\n",
    "import tools.scDEED as scd\n",
    "import tools.clustering_opt as co\n",
    "import tools.util_plot as nmd_plot\n",
    "import tools.util_imputation as nmd_imp\n",
    "\n",
    "import fi_nomad as nmd\n",
    "from fi_nomad.types import kernelInputTypes\n",
    "from fi_nomad.types import KernelStrategy\n",
    "from fi_nomad.types import InitializationStrategy\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from scipy.stats import kendalltau\n",
    "import importlib\n",
    "import logging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# logging has to manually turned on to see nomad output\n",
    "logging.basicConfig(level=logging.INFO)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = \"../../data/S3\"\n",
    "figure_path = f\"{data_path}/figures/imputation\"\n",
    "layer = \"counts\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1544 × 5553\n",
       "    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'outlier', 'n_genes', 'total_counts_norm', 'total_counts_scale', 'embedding_reliability', 'reliability_score', 'null_reliability_score', 'leiden_opt_PCA'\n",
       "    var: 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'var_counts', 'is_scd_outlier', 'nb_overdisp', 'nb_overdisp_cutoff', 'nb_mean', 'nb_umi', 'Intercept_step1_sct', 'log_umi_step1_sct', 'dispersion_step1_sct', 'mean', 'std'\n",
       "    uns: 'BacSC_params_PCA', 'PCA', 'leiden', 'leiden_opt_PCA_colors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'normalized_counts', 'vst_counts'\n",
       "    obsp: 'PCA_connectivities', 'PCA_distances'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_counts = sc.read_h5ad(data_path + \"/filtered_data_maxpool_processed_and_pca.h5ad\")\n",
    "data_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read expression level class data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>tpm</th>\n",
       "      <th>expression_level</th>\n",
       "      <th>log_tpm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA0001</th>\n",
       "      <td>dnaA</td>\n",
       "      <td>333.097801</td>\n",
       "      <td>high</td>\n",
       "      <td>8.379802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0002</th>\n",
       "      <td>dnaN</td>\n",
       "      <td>299.127321</td>\n",
       "      <td>high</td>\n",
       "      <td>8.224616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0003</th>\n",
       "      <td>recF</td>\n",
       "      <td>128.151975</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.001712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0004</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>138.579052</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.114565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0005</th>\n",
       "      <td>lptA</td>\n",
       "      <td>141.413217</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.143773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5566</th>\n",
       "      <td>PA5566</td>\n",
       "      <td>45.077831</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.494346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5567</th>\n",
       "      <td>PA5567</td>\n",
       "      <td>49.149167</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.619095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5568</th>\n",
       "      <td>PA5568</td>\n",
       "      <td>208.178186</td>\n",
       "      <td>high</td>\n",
       "      <td>7.701675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5569</th>\n",
       "      <td>rnpA</td>\n",
       "      <td>3387.343294</td>\n",
       "      <td>high</td>\n",
       "      <td>11.725938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5570</th>\n",
       "      <td>rpmH</td>\n",
       "      <td>4824.078870</td>\n",
       "      <td>high</td>\n",
       "      <td>12.236038</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Name          tpm expression_level    log_tpm\n",
       "locus_tag                                                 \n",
       "PA0001       dnaA   333.097801             high   8.379802\n",
       "PA0002       dnaN   299.127321             high   8.224616\n",
       "PA0003       recF   128.151975      medium_high   7.001712\n",
       "PA0004       gyrB   138.579052      medium_high   7.114565\n",
       "PA0005       lptA   141.413217      medium_high   7.143773\n",
       "...           ...          ...              ...        ...\n",
       "PA5566     PA5566    45.077831       medium_low   5.494346\n",
       "PA5567     PA5567    49.149167       medium_low   5.619095\n",
       "PA5568     PA5568   208.178186             high   7.701675\n",
       "PA5569       rnpA  3387.343294             high  11.725938\n",
       "PA5570       rpmH  4824.078870             high  12.236038\n",
       "\n",
       "[5572 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_class = pd.read_csv(\n",
    "    \"../../data/control_expression_classification.tsv\", delimiter=\"\\t\"\n",
    ")\n",
    "expr_class.set_index(keys=\"locus_tag\", inplace=True)\n",
    "expr_class"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- `mean_counts`: mean gene expression\n",
    "- `n_cells`: number of cells that express gene (so (n - `n_cells`) / n = sparsity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>log1p_mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>var_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>nb_umi</th>\n",
       "      <th>Intercept_step1_sct</th>\n",
       "      <th>log_umi_step1_sct</th>\n",
       "      <th>dispersion_step1_sct</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>Name</th>\n",
       "      <th>tpm</th>\n",
       "      <th>expression_level</th>\n",
       "      <th>log_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA0001</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>735</td>\n",
       "      <td>0.666451</td>\n",
       "      <td>0.510696</td>\n",
       "      <td>52.396373</td>\n",
       "      <td>1029.0</td>\n",
       "      <td>6.937314</td>\n",
       "      <td>735</td>\n",
       "      <td>0.755973</td>\n",
       "      <td>...</td>\n",
       "      <td>2.182498</td>\n",
       "      <td>-6.941972</td>\n",
       "      <td>2.135988</td>\n",
       "      <td>0.009296</td>\n",
       "      <td>1.608654</td>\n",
       "      <td>0.117757</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>333.097801</td>\n",
       "      <td>high</td>\n",
       "      <td>8.379802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0002</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>541</td>\n",
       "      <td>0.468912</td>\n",
       "      <td>0.384522</td>\n",
       "      <td>64.961140</td>\n",
       "      <td>724.0</td>\n",
       "      <td>6.586172</td>\n",
       "      <td>541</td>\n",
       "      <td>0.588412</td>\n",
       "      <td>...</td>\n",
       "      <td>2.249176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.229464</td>\n",
       "      <td>0.142842</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>299.127321</td>\n",
       "      <td>high</td>\n",
       "      <td>8.224616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0003</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>890</td>\n",
       "      <td>0.893782</td>\n",
       "      <td>0.638576</td>\n",
       "      <td>42.357513</td>\n",
       "      <td>1380.0</td>\n",
       "      <td>7.230563</td>\n",
       "      <td>890</td>\n",
       "      <td>1.043122</td>\n",
       "      <td>...</td>\n",
       "      <td>2.153516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.691644</td>\n",
       "      <td>0.124664</td>\n",
       "      <td>recF</td>\n",
       "      <td>128.151975</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.001712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0004</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>463</td>\n",
       "      <td>0.373705</td>\n",
       "      <td>0.317511</td>\n",
       "      <td>70.012953</td>\n",
       "      <td>577.0</td>\n",
       "      <td>6.359574</td>\n",
       "      <td>463</td>\n",
       "      <td>0.398557</td>\n",
       "      <td>...</td>\n",
       "      <td>2.293357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.974741</td>\n",
       "      <td>0.150374</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>138.579052</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.114565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA0005</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>173</td>\n",
       "      <td>0.120466</td>\n",
       "      <td>0.113745</td>\n",
       "      <td>88.795337</td>\n",
       "      <td>186.0</td>\n",
       "      <td>5.231109</td>\n",
       "      <td>173</td>\n",
       "      <td>0.124089</td>\n",
       "      <td>...</td>\n",
       "      <td>2.494667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.471332</td>\n",
       "      <td>0.257211</td>\n",
       "      <td>lptA</td>\n",
       "      <td>141.413217</td>\n",
       "      <td>medium_high</td>\n",
       "      <td>7.143773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5566</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>14</td>\n",
       "      <td>0.009715</td>\n",
       "      <td>0.009668</td>\n",
       "      <td>99.093264</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.772589</td>\n",
       "      <td>14</td>\n",
       "      <td>0.010916</td>\n",
       "      <td>...</td>\n",
       "      <td>2.336949</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.570001</td>\n",
       "      <td>0.321510</td>\n",
       "      <td>PA5566</td>\n",
       "      <td>45.077831</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.494346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5567</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>163</td>\n",
       "      <td>0.126943</td>\n",
       "      <td>0.119509</td>\n",
       "      <td>89.443005</td>\n",
       "      <td>196.0</td>\n",
       "      <td>5.283204</td>\n",
       "      <td>163</td>\n",
       "      <td>0.161347</td>\n",
       "      <td>...</td>\n",
       "      <td>2.493876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.463125</td>\n",
       "      <td>0.270051</td>\n",
       "      <td>PA5567</td>\n",
       "      <td>49.149167</td>\n",
       "      <td>medium_low</td>\n",
       "      <td>5.619095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5568</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>740</td>\n",
       "      <td>0.659974</td>\n",
       "      <td>0.506802</td>\n",
       "      <td>52.072539</td>\n",
       "      <td>1019.0</td>\n",
       "      <td>6.927558</td>\n",
       "      <td>740</td>\n",
       "      <td>0.750315</td>\n",
       "      <td>...</td>\n",
       "      <td>2.183110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.604845</td>\n",
       "      <td>0.117774</td>\n",
       "      <td>PA5568</td>\n",
       "      <td>208.178186</td>\n",
       "      <td>high</td>\n",
       "      <td>7.701675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5569</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>1430</td>\n",
       "      <td>3.498057</td>\n",
       "      <td>1.503646</td>\n",
       "      <td>7.383420</td>\n",
       "      <td>5401.0</td>\n",
       "      <td>8.594525</td>\n",
       "      <td>1430</td>\n",
       "      <td>14.170981</td>\n",
       "      <td>...</td>\n",
       "      <td>2.223133</td>\n",
       "      <td>-6.188664</td>\n",
       "      <td>2.421585</td>\n",
       "      <td>0.057214</td>\n",
       "      <td>1.302429</td>\n",
       "      <td>0.374002</td>\n",
       "      <td>rnpA</td>\n",
       "      <td>3387.343294</td>\n",
       "      <td>high</td>\n",
       "      <td>11.725938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA5570</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>PA01</td>\n",
       "      <td>480</td>\n",
       "      <td>0.466969</td>\n",
       "      <td>0.383198</td>\n",
       "      <td>68.911917</td>\n",
       "      <td>721.0</td>\n",
       "      <td>6.582025</td>\n",
       "      <td>480</td>\n",
       "      <td>0.746318</td>\n",
       "      <td>...</td>\n",
       "      <td>2.262201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.157867</td>\n",
       "      <td>0.157798</td>\n",
       "      <td>rpmH</td>\n",
       "      <td>4824.078870</td>\n",
       "      <td>high</td>\n",
       "      <td>12.236038</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5553 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          feature_types genome  n_cells_by_counts  mean_counts  \\\n",
       "PA0001  Gene Expression   PA01                735     0.666451   \n",
       "PA0002  Gene Expression   PA01                541     0.468912   \n",
       "PA0003  Gene Expression   PA01                890     0.893782   \n",
       "PA0004  Gene Expression   PA01                463     0.373705   \n",
       "PA0005  Gene Expression   PA01                173     0.120466   \n",
       "...                 ...    ...                ...          ...   \n",
       "PA5566  Gene Expression   PA01                 14     0.009715   \n",
       "PA5567  Gene Expression   PA01                163     0.126943   \n",
       "PA5568  Gene Expression   PA01                740     0.659974   \n",
       "PA5569  Gene Expression   PA01               1430     3.498057   \n",
       "PA5570  Gene Expression   PA01                480     0.466969   \n",
       "\n",
       "        log1p_mean_counts  pct_dropout_by_counts  total_counts  \\\n",
       "PA0001           0.510696              52.396373        1029.0   \n",
       "PA0002           0.384522              64.961140         724.0   \n",
       "PA0003           0.638576              42.357513        1380.0   \n",
       "PA0004           0.317511              70.012953         577.0   \n",
       "PA0005           0.113745              88.795337         186.0   \n",
       "...                   ...                    ...           ...   \n",
       "PA5566           0.009668              99.093264          15.0   \n",
       "PA5567           0.119509              89.443005         196.0   \n",
       "PA5568           0.506802              52.072539        1019.0   \n",
       "PA5569           1.503646               7.383420        5401.0   \n",
       "PA5570           0.383198              68.911917         721.0   \n",
       "\n",
       "        log1p_total_counts  n_cells  var_counts  ...    nb_umi  \\\n",
       "PA0001            6.937314      735    0.755973  ...  2.182498   \n",
       "PA0002            6.586172      541    0.588412  ...  2.249176   \n",
       "PA0003            7.230563      890    1.043122  ...  2.153516   \n",
       "PA0004            6.359574      463    0.398557  ...  2.293357   \n",
       "PA0005            5.231109      173    0.124089  ...  2.494667   \n",
       "...                    ...      ...         ...  ...       ...   \n",
       "PA5566            2.772589       14    0.010916  ...  2.336949   \n",
       "PA5567            5.283204      163    0.161347  ...  2.493876   \n",
       "PA5568            6.927558      740    0.750315  ...  2.183110   \n",
       "PA5569            8.594525     1430   14.170981  ...  2.223133   \n",
       "PA5570            6.582025      480    0.746318  ...  2.262201   \n",
       "\n",
       "        Intercept_step1_sct  log_umi_step1_sct  dispersion_step1_sct  \\\n",
       "PA0001            -6.941972           2.135988              0.009296   \n",
       "PA0002                  NaN                NaN                   NaN   \n",
       "PA0003                  NaN                NaN                   NaN   \n",
       "PA0004                  NaN                NaN                   NaN   \n",
       "PA0005                  NaN                NaN                   NaN   \n",
       "...                     ...                ...                   ...   \n",
       "PA5566                  NaN                NaN                   NaN   \n",
       "PA5567                  NaN                NaN                   NaN   \n",
       "PA5568                  NaN                NaN                   NaN   \n",
       "PA5569            -6.188664           2.421585              0.057214   \n",
       "PA5570                  NaN                NaN                   NaN   \n",
       "\n",
       "            mean       std    Name          tpm  expression_level    log_tpm  \n",
       "PA0001  1.608654  0.117757    dnaA   333.097801              high   8.379802  \n",
       "PA0002  1.229464  0.142842    dnaN   299.127321              high   8.224616  \n",
       "PA0003  1.691644  0.124664    recF   128.151975       medium_high   7.001712  \n",
       "PA0004  0.974741  0.150374    gyrB   138.579052       medium_high   7.114565  \n",
       "PA0005 -0.471332  0.257211    lptA   141.413217       medium_high   7.143773  \n",
       "...          ...       ...     ...          ...               ...        ...  \n",
       "PA5566 -3.570001  0.321510  PA5566    45.077831        medium_low   5.494346  \n",
       "PA5567 -0.463125  0.270051  PA5567    49.149167        medium_low   5.619095  \n",
       "PA5568  1.604845  0.117774  PA5568   208.178186              high   7.701675  \n",
       "PA5569  1.302429  0.374002    rnpA  3387.343294              high  11.725938  \n",
       "PA5570  1.157867  0.157798    rpmH  4824.078870              high  12.236038  \n",
       "\n",
       "[5553 rows x 24 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_counts.var = data_counts.var.join(expr_class)\n",
    "data_counts.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit NMD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "latent_rank = 10\n",
    "momentum_beta = 0.7\n",
    "max_iterations = 1000\n",
    "\n",
    "init_strat = InitializationStrategy.ROWWISE_MEAN\n",
    "kernel_strat = KernelStrategy.MOMENTUM_3_BLOCK_MODEL_FREE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "m, n = data_counts.X.shape\n",
    "X_nmd = ut.convert_to_dense(data_counts, layer=layer)\n",
    "X_nmd = X_nmd.astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:fi_nomad.entry:\tInitiating run, target_rank: 10, tolerance: None\n",
      "INFO:fi_nomad.entry:1000 total, final loss Not Tracked\n",
      "INFO:fi_nomad.entry:\tInitialization took 32.63490094896406 loop took 230.56548640300753 overall (0.23056548640300753/ea)\n"
     ]
    }
   ],
   "source": [
    "result = nmd.decompose(\n",
    "    X_nmd,\n",
    "    latent_rank,\n",
    "    kernel_strategy=kernel_strat,\n",
    "    initialization=init_strat,\n",
    "    kernel_params=kernelInputTypes.Momentum3BlockAdditionalParameters(\n",
    "        momentum_beta=momentum_beta\n",
    "    ),\n",
    "    manual_max_iterations=max_iterations,\n",
    "    verbose=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_counts.obsm[\"X_nmd\"] = result.factors[0]\n",
    "data_counts.varm[\"NMD_components\"] = result.factors[1].T\n",
    "data_counts.layers[\"Theta\"] = result.factors[0] @ result.factors[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output file for pretty plotting in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(data_counts.layers[\"Theta\"], columns=data_counts.var_names).to_csv(\n",
    "    f\"{data_path}/Theta_3BNMD_r{latent_rank}.csv\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample gene pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'tools.util_imputation' from '/Users/stffn/projects/thesis_dev/msc_thesis/application/tools/util_imputation.py'>"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(nmd_imp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes = 50  # number of gene pairs\n",
    "expression_levels_2_compare = [\"low\", \"medium_high\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>low</th>\n",
       "      <th>medium_high</th>\n",
       "      <th>dist</th>\n",
       "      <th>gene_pair_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA0688</td>\n",
       "      <td>PA0109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA2908</td>\n",
       "      <td>PA4791</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA4293</td>\n",
       "      <td>PA0176</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA3760</td>\n",
       "      <td>PA0108</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA1227</td>\n",
       "      <td>PA3205</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      low medium_high      dist  gene_pair_id\n",
       "0  PA0688      PA0109  0.000000             0\n",
       "1  PA2908      PA4791  0.000012             1\n",
       "2  PA4293      PA0176  0.000012             2\n",
       "3  PA3760      PA0108  0.000104             3\n",
       "4  PA1227      PA3205  0.000012             4"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pairs = nmd_imp.sample_gene_pairs(\n",
    "    data_counts,\n",
    "    n=n_genes,\n",
    "    measure_var=[\"mean_counts\", \"pct_dropout_by_counts\"],\n",
    "    expr_lvls=expression_levels_2_compare,\n",
    ")\n",
    "gene_pairs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get all negative Theta values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "      <th>gene_pair_id</th>\n",
       "      <th>dist</th>\n",
       "      <th>expression_lvl_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA0688</td>\n",
       "      <td>0.005996</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA0688</td>\n",
       "      <td>0.011742</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA0688</td>\n",
       "      <td>0.010352</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA0688</td>\n",
       "      <td>0.022282</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA0688</td>\n",
       "      <td>0.019896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene  Expression  gene_pair_id  dist expression_lvl_class\n",
       "0  PA0688    0.005996             0   0.0                  low\n",
       "1  PA0688    0.011742             0   0.0                  low\n",
       "2  PA0688    0.010352             0   0.0                  low\n",
       "3  PA0688    0.022282             0   0.0                  low\n",
       "4  PA0688    0.019896             0   0.0                  low"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_df = nmd_imp.make_gene_pair_df(gene_pairs, adata=data_counts)\n",
    "gene_pair_df.sort_values(by=[\"gene_pair_id\", \"dist\"]).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export to csv for visualization in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_df.to_csv(\n",
    "    f\"{data_path}/gene_pair_df_4_viz_{n_genes}genes_3BNMD_rank{latent_rank}_{expression_levels_2_compare[0]}-{expression_levels_2_compare[1]}_mean_and_sparsity_whereXzero.csv\",\n",
    "    index=False,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bacsc-jupyter-lrz",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
